system/Security/Security.php

```diff
@@ -329,7 +340,33 @@ private function getPostedToken(RequestInterface $request): ?string
      */
     public function getHash(): ?string
     {
-        return $this->hash;
+        return $this->tokenRandomize ? $this->randomize($this->hash) : $this->hash;
+    }
+
+    /**
+     * Randomize hash to avoid BREACH attacks.
+     */
+    protected function randomize(string $hash): string
+    {
+        $key   = random_bytes(self::CSRF_HASH_BYTES);
+        $value = hex2bin($hash);
+
+        if ($value === false) {
+            throw new LogicException('$hash is invalid: ' . $hash);
```

Comment:
Hash in this context I assume is a binary string. So outputting the full text would IMO not dev friendly. I think we can forego outputting the $hash

Reply:
I think `$hash` is still the hex (or supposed to be) and `$value` is the binary, so this should be okay.

Reply:
Yes, `$hash` is the CSRF hash that is hex.
But I rename variable name `$value` not to be confused.

Reply:
Indeed it is hex.